components/custom_board/tas5805m/tas5805m.c

```diff
@@ -0,0 +1,277 @@
+    .audio_codec_get_volume = tas5805m_get_volume,
+    .audio_hal_lock = NULL,
+    .handle = NULL,
+};
+
+/* Init the I2C Driver */
+
+void i2c_master_init()
+{
+  int i2c_master_port = I2C_MASTER_NUM;
+
+  get_i2c_pins(I2C_NUM_0, &i2c_cfg);
+
+  esp_err_t res = i2c_param_config(i2c_master_port, &i2c_cfg);
+  printf("Driver param setup : %d\n", res);
+  res = i2c_driver_install(i2c_master_port, i2c_cfg.mode,
+                           I2C_MASTER_RX_BUF_DISABLE, I2C_MASTER_TX_BUF_DISABLE,
+                           0);
+  printf("Driver installed   : %d\n", res);
+}
+
+/* Helper Functions */
+
+// Mapping between Value Areas implementation from arduino.cc
+long map(long x, long in_min, long in_max, long out_min, long out_max)
+{
+  return (x - in_min) * (out_max - out_min) / (in_max - in_min) + out_min;
+}
+// Reading of TAS5805M-Register
+
+esp_err_t tas5805m_read_byte(uint8_t register_name, uint8_t *data)
+{
+
+  int ret;
+  i2c_cmd_handle_t cmd = i2c_cmd_link_create();
+  i2c_master_start(cmd);
+  i2c_master_write_byte(cmd, TAS5805M_ADDRESS << 1 | WRITE_BIT, ACK_CHECK_EN);
+  i2c_master_write_byte(cmd, register_name, ACK_CHECK_EN);
+  i2c_master_stop(cmd);
+  ret = i2c_master_cmd_begin(I2C_TAS5805M_MASTER_NUM, cmd, 1000 / portTICK_RATE_MS);
+  i2c_cmd_link_delete(cmd);
+
+  if (ret != ESP_OK)
+  {
+    ESP_LOGW(TAG, "I2C ERROR");
+  }
+
+  vTaskDelay(1 / portTICK_RATE_MS);
+  cmd = i2c_cmd_link_create();
+  i2c_master_start(cmd);
+  i2c_master_write_byte(cmd, TAS5805M_ADDRESS << 1 | READ_BIT, ACK_CHECK_EN);
+  i2c_master_read_byte(cmd, data, NACK_VAL);
+  i2c_master_stop(cmd);
+  ret = i2c_master_cmd_begin(I2C_TAS5805M_MASTER_NUM, cmd, 1000 / portTICK_RATE_MS);
+  i2c_cmd_link_delete(cmd);
+
+  return ret;
+}
+
+// Writing of TAS5805M-Register
+
+esp_err_t tas5805m_write_byte(uint8_t register_name, uint8_t value)
+{
+  int ret;
+  i2c_cmd_handle_t cmd = i2c_cmd_link_create();
+  i2c_master_start(cmd);
+  i2c_master_write_byte(cmd, TAS5805M_ADDRESS << 1 | WRITE_BIT, ACK_CHECK_EN);
+  i2c_master_write_byte(cmd, register_name, ACK_CHECK_EN);
+  i2c_master_write_byte(cmd, value, ACK_CHECK_EN);
+  i2c_master_stop(cmd);
+  ret = i2c_master_cmd_begin(I2C_TAS5805M_MASTER_NUM, cmd, 1000 / portTICK_RATE_MS);
+  i2c_cmd_link_delete(cmd);
+
+  return ret;
+}
+
+// Inits the TAS5805M change Settings in Menuconfig to enable Bridge-Mode
+
+esp_err_t tas5805m_init()
+{
+  int ret;
+  // Init the I2C-Driver
+  i2c_master_init();
+  /* Register the PDN pin as output and write 1 to enable the TAS chip */
+  /* TAS5805M.INIT() */
+  gpio_config_t io_conf;
+  io_conf.intr_type = GPIO_INTR_DISABLE;
+  io_conf.mode = GPIO_MODE_OUTPUT;
+  io_conf.pin_bit_mask = TAS5805M_GPIO_PDN_MASK;
+  io_conf.pull_down_en = GPIO_PULLDOWN_DISABLE;
+  io_conf.pull_up_en = GPIO_PULLUP_DISABLE;
+  ESP_LOGW(TAG, "Power down pin: %d", TAS5805M_GPIO_PDN);
+  gpio_config(&io_conf);
+  gpio_set_level(TAS5805M_GPIO_PDN, 0);
+  vTaskDelay(200 / portTICK_RATE_MS);
```

Comment:
Is this really needed seems quite long

Reply:
@DerPicknicker do these delays really need to be 100s of ms? Shouldn't these be around a couple us?